src/librustc_resolve/diagnostics.rs

```diff
@@ -354,6 +354,17 @@ impl<'a> Resolver<'a> {
                     span, "defaulted type parameters cannot be forward declared".to_string());
                 err
             }
+            ResolutionError::SelfInTyParamDefault => {
+                let mut err = struct_span_err!(
+                    self.session,
+                    span,
+                    E0735,
+                    "type parameters with a default cannot use `Self`"
```

Comment:
I prefer the wording in the documentation.
```suggestion
                    "type parameters cannot use `Self` in their defaults"
```